src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSLambdaHttpUtils.cs

```diff
@@ -0,0 +1,135 @@
+
+            switch (input)
+            {
+                case APIGatewayProxyRequest request:
+                    httpScheme = AWSLambdaUtils.GetHeaderValues(request, HeaderXForwardedProto)?.FirstOrDefault();
+                    httpTarget = string.Concat(request.Path ?? string.Empty, GetQueryString(request));
+                    httpMethod = request.HttpMethod;
+                    var hostHeader = AWSLambdaUtils.GetHeaderValues(request, HeaderHost)?.FirstOrDefault();
+                    (hostName, hostPort) = GetHostAndPort(httpScheme, hostHeader);
+                    break;
+                case APIGatewayHttpApiV2ProxyRequest requestV2:
+                    httpScheme = AWSLambdaUtils.GetHeaderValues(requestV2, HeaderXForwardedProto)?.FirstOrDefault();
+                    httpTarget = string.Concat(requestV2?.RawPath ?? string.Empty, GetQueryString(requestV2));
+                    httpMethod = requestV2?.RequestContext?.Http?.Method;
+                    var hostHeaderV2 = AWSLambdaUtils.GetHeaderValues(requestV2, HeaderHost)?.FirstOrDefault();
+                    (hostName, hostPort) = GetHostAndPort(httpScheme, hostHeaderV2);
+                    break;
+            }
+
+            tags.AddTagIfNotNull(SemanticConventions.AttributeHttpScheme, httpScheme);
+            tags.AddTagIfNotNull(SemanticConventions.AttributeHttpTarget, httpTarget);
+            tags.AddTagIfNotNull(SemanticConventions.AttributeHttpMethod, httpMethod);
+            tags.AddTagIfNotNull(SemanticConventions.AttributeNetHostName, hostName);
+            tags.AddTagIfNotNull(SemanticConventions.AttributeNetHostPort, hostPort);
+
+            return tags;
+        }
+
+        internal static void SetHttpTagsFromResult(Activity activity, object result)
+        {
+            switch (result)
+            {
+                case APIGatewayProxyResponse response:
+                    activity.SetTag(SemanticConventions.AttributeHttpStatusCode, response.StatusCode);
+                    break;
+                case APIGatewayHttpApiV2ProxyResponse responseV2:
+                    activity.SetTag(SemanticConventions.AttributeHttpStatusCode, responseV2.StatusCode);
+                    break;
+            }
+        }
+
+        internal static string GetQueryString(APIGatewayProxyRequest request)
+        {
+            if (request.MultiValueQueryStringParameters == null)
+            {
+                return string.Empty;
+            }
+
+            var queryString = new StringBuilder();
+            var separator = '?';
+            foreach (var parameterKvp in request.MultiValueQueryStringParameters)
+            {
+                // Multiple values for the same parameter will be added to query
+                // as ampersand separated: name=value1&name=value2
+                foreach (var value in parameterKvp.Value)
+                {
+                    queryString.Append(string.Concat(separator, parameterKvp.Key, "=", value));
```

Comment:
If we are using a string builder, I think we should rely on it for concatenation and not allocate additional string. It looks like there is even an overload for interpolated strings <https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.append?view=net-6.0#system-text-stringbuilder-append(system-text-stringbuilder-appendinterpolatedstringhandler@)>

So this should be among the most efficient ways:

```suggestion
                    queryString.Append($"{separator}{parameterKvp.Key}={value}");
```

As the string builder will internally evaluate the $-string by calling Append multiple times. See here: https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/stringbuilder-append-evaluation-order#new-behavior

Reply:
Sorry, on second thought, while this is a very cool feature, it requires not only C# 10, but also .NET 6 (which we can't rely on). So it would probably be better to spell out the `Append` manually.

Reply:
Done